$OpenBSD: patch-property_c,v 1.9 2014/12/16 11:16:05 dcoppa Exp $

commit ed3d0489ced8c4f56d2af990979ba75cfa5107b8
Author: Uli Schlachter <psychon@znc.in>
Date:   Sun Dec 7 14:09:35 2014 +0100

Implement icon_pixmap and icon_mask from WM_HINTS (FS#1297)

Fun fact: ICCCM specifies that icon_pixmap must have depth 1. Xterm uses a
pixmap with depth 24. Yay... As such, I don't have any test for the depth == 1
case and will just assume that it does the right thing. If it doesn't, I bet no
one will notice anyway.

Signed-off-by: Uli Schlachter <psychon@znc.in>

--- property.c.orig	Fri Apr 11 04:07:10 2014
+++ property.c	Tue Dec 16 04:47:13 2014
@@ -205,6 +205,19 @@ property_update_wm_hints(client_t *c, xcb_get_property
     if(wmh.flags & XCB_ICCCM_WM_HINT_WINDOW_GROUP)
         client_set_group_window(globalconf.L, -1, wmh.window_group);
 
+    if(!c->have_ewmh_icon)
+    {
+        if(wmh.flags & XCB_ICCCM_WM_HINT_ICON_PIXMAP)
+        {
+            if(wmh.flags & XCB_ICCCM_WM_HINT_ICON_MASK)
+                client_set_icon_from_pixmaps(c, wmh.icon_pixmap, wmh.icon_mask);
+            else
+                client_set_icon_from_pixmaps(c, wmh.icon_pixmap, XCB_NONE);
+        }
+        else
+            client_set_icon(c, NULL);
+    }
+
     lua_pop(globalconf.L, 1);
 }
 
@@ -261,6 +274,7 @@ property_update_net_wm_icon(client_t *c, xcb_get_prope
     if(!surface)
         return;
 
+    c->have_ewmh_icon = true;
     client_set_icon(c, surface);
     cairo_surface_destroy(surface);
 }
